server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java

```diff
@@ -14,451 +14,506 @@
-        }).when(channelManager).sendRequest(any(AssignReplicasToDirsRequest.Builder.class),
-                any(ControllerRequestCompletionHandler.class));
-
-        for (int i = 0; i < 300; i++) {
-            manager.onAssignment(new TopicIdPartition(TOPIC_1, i % 5), DIR_1, "testOnCompletion", readyToAssert::countDown);
+    public void testStartAndShutdown() throws Exception {
+        try (TestEnv testEnv = new TestEnv()) {
         }
+    }
 
-        TestUtils.waitForCondition(() -> {
-            time.sleep(TimeUnit.SECONDS.toMillis(1));
-            manager.wakeup();
-            return readyToAssert.await(1, TimeUnit.MILLISECONDS);
-        }, "Timed out waiting for AssignReplicasToDirsRequest to be sent.");
+    @Test
+    public void testSuccessfulAssignment() throws Exception {
+        try (TestEnv testEnv = new TestEnv()) {
+            assertEquals(0, testEnv.queuedReplicaToDirAssignments());
+            testEnv.onAssignment(new TopicIdPartition(TOPIC_1, 0), DIR_1);
+            TestUtils.retryOnExceptionWithTimeout(60_000, () -> {
+                assertEquals(1, testEnv.assignmentsManager.numPending());
+                assertEquals(1, testEnv.queuedReplicaToDirAssignments());
+            });
+            assertEquals(0, testEnv.assignmentsManager.previousGlobalFailures());
+            assertEquals(1, testEnv.assignmentsManager.numInFlight());
+            testEnv.successfullyCompleteCallbackOfRequestAssigningTopic1ToDir1();
+            TestUtils.retryOnExceptionWithTimeout(60_000, () -> {
+                assertEquals(0, testEnv.assignmentsManager.numPending());
+                assertEquals(0, testEnv.queuedReplicaToDirAssignments());
+                assertEquals(1, testEnv.success(new TopicIdPartition(TOPIC_1, 0)));
+            });
+            assertEquals(0, testEnv.assignmentsManager.previousGlobalFailures());
+        }
     }
 
-    private static ClientResponse buildSuccessfulResponse(AssignReplicasToDirsRequestData request) {
-        return buildResponse(request, topicIdPartition -> Errors.NONE);
+    @ParameterizedTest
+    @ValueSource(strings = {"invalidRequest", "disconnection"})
+    public void testUnSuccessfulRequestCausesRetransmission(String failureType) throws Exception {
+        try (TestEnv testEnv = new TestEnv()) {
+            testEnv.onAssignment(new TopicIdPartition(TOPIC_1, 0), DIR_1);
+            TestUtils.retryOnExceptionWithTimeout(60_000, () -> {
+                assertEquals(1, testEnv.assignmentsManager.numPending());
+            });
+            if (failureType.equals("invalidRequest")) {
+                testEnv.channelManager.completeCallback(req -> {
+                    return mockClientResponse(new AssignReplicasToDirsResponseData().
+                        setErrorCode(Errors.INVALID_REQUEST.code()));
+                });
+            } else if (failureType.equals("disconnection")) {
+                testEnv.channelManager.completeCallback(req -> Optional.empty());
```

Comment:
This is testing the timeout path, not the disconnection.

Reply:
Yeah. I'll change the failureType string to "timeout".